test/unit/puppetlabs/services/legacy_routes/legacy_routes_core_test.clj

```diff
@@ -0,0 +1,145 @@
+  (let [handler (assertion-wrapper identity assertion)]
+    (build-ring-handler
+      handler master-mount master-api-version
+      identity ca-mount ca-api-version)))
+
+(defn test-munged-request
+  "Validate a request by threading it through the legacy-routes request handler
+  into an assertion-wrapper ring middleware."
+  [request assertion]
+  ((validating-app assertion) request))
+
+(deftest test-legacy-routes
+  (let [app     (build-ring-handler
+                  identity master-mount master-api-version
+                  identity ca-mount ca-api-version)
+        request (fn r ([path] (r :get path))
+                      ([method path]
+                       (app (mock/request method path))))
+        environment "production"
+        route-val "foo"]
+    (is (= 200 (:status (request :get "/v2.0/environments"))))
+    (is (= 404 (:status (request :get "/does/not/exist"))))
+
+    (testing "Legacy master routes"
+      (doseq [[method paths] {:get ["node"
+                                    "facts"
+                                    "file_content"
+                                    "file_metadatas"
+                                    "file_metadata"
+                                    "file_bucket_file"
+                                    "catalog"
+                                    "resource_type"
+                                    "resource_types"
+                                    "facts_search"]
+                              :put ["report"]
+                              :head ["file_bucket_file"]
+                              :post ["catalog"]}
+              path paths]
+        (let [resp (request method (str "/" environment "/" path "/" route-val))]
+          (is (= 200 (:status resp)))
+          (is (= (str master-mount "/" master-api-version "/" path "/" route-val)
+                 (:uri resp)))
+          (is (= (ring-codec/form-encode {:environment environment})
+                 (:query-string resp))))))
+
+    (testing "Legacy ca routes"
+      (doseq [path ["certificate_status"
+                    "certificate_statuses"
+                    "certificate"
+                    "certificate_revocation_list"
+                    "certificate_request"
+                    "certificate_requests"]]
+        (let [resp (request (str "/" environment "/" path "/" route-val))]
+          (is (= 200 (:status resp)))
+          (is (= (str ca-mount "/" ca-api-version "/" path "/" route-val)))
+          (is (= (ring-codec/form-encode {:environment environment}))))))))
+
+(deftest test-v3-header-munging
+  (testing "(SERVER-548) Header munging"
+    (testing "Accept: raw is munged to Accept: binary for all paths"
+      (doseq [[path methods] accept-header-test-data
+              [method examples] methods
+              [example expected] examples
+              :let [mockreq (mock/request method path)
+                    request (assoc-in mockreq accept-header example)]]
+        (test-munged-request request
```

Comment:
It seems like another approach that would work for this would be to just do the expected/actual test against what is returned from `test-munged-request`.  That would have the headers in the Ring app as they were after munging from all of the other middleware - and might avoid the need to pass the assertion function all the way through.  I wouldn't hold up this PR to get that changed, but it seems like that approach might be a little easier to read.